_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,274 @@
++  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"completedOn" format:"date-time"`
++  Status      TestResultStatus  `gorm:"not null" enums:"Failed,Passed,Skipped"`
++}
+```
+
+#### New functions, methods, and structs
+
+```go
+// artifact.go
+type File struct {
+  name string
+  fileName string
+  data []bytes
+}
+// @router /build/{buildid}/test-results [put]
+func (m ArtifactModule) putTestResultsHandler(c *gin.Context) {
+  // parseMultipartFormData to get the files
+  // for each file, parse and store in db
+  // then sends a response with the summary
+}
+// @router /build/{buildid]/test-result-details [get]
+func (m ArtifactModule) getBuildAllTestResultDetailsHandler(c *gin.Context) {
+  // fetch all test result details for build
+}
+// @router /build/{buildid}/test-result-details/{testresultsummaryid} [get]
+func (m ArtifactModule) getBuildTestResultDetailsHandler(c *gin.Context) {
+  // fetch all test result details for specific test for build
+}
+// @router /build/{buildid}/test-results-summary [get]
+func (m ArtifactModule) getBuildTestResultsSummary(c *gin.Context) {
+  // fetch sum of each artifact's test result summary for build
+  // as well as the list of each artifact's test result summary
+}
+
+func parseMultipartFormData(c *gin.Context) []*File {
+  // ...
+}
+
+func parseTRX(file *File) []TestResult, TestResultSummary {
+  // ...
+}
+```
+
+The new endpoint `GET /build/{buildid}/test-results-summary` fetches
+all test result summaries for a build and constructs a JSON response like:
+
+`GET /build/42/test-results-summary`
+
+```json
+{
+  "failed": 4,
+  "passed": 5,
+  "skipped": 0,
+  "artifacts": [
+    {
+      "artifactId": 123,
+      "filename": "MyResults1.trx",
+      "failed": 1,
+      "passed": 3,
+      "skipped": 0
+    },
+    {
+      "artifactId": 124,
+      "filename": "MyResults2.trx",
+      "failed": 3,
+      "passed": 2,
+      "skipped": 0
+    }
+  ]
+}
+```
+
+### wharf-web
+
+wharf-web changes to use the new endpoints
+
+- `GET /build/{buildid}/test-results-summary`\
+  To get summary of all test result summaries for build.
+
+- `GET /build/{buildid}/test-result-details`\
+  To get all test result details for build.
+  Supports pagination: `?limit=10&offset=20&orderby=asc`
+
+- `GET /build/{buildid}/test-result-details/{testresultid}`\
+  To get specific test result details for build.
+
+- `POST /build/{buildid}/test-results`\
+  To upload test result files for build.\
+  Also responds with the same data as if calling `GET /build/{buildid}/test-results-summary` after.
+
+Deprecated endpoint (Add @deprecated flag in wharf-api)
+
+- `GET /build/{buildid}/tests-results`
+
+There would also be a way to view a build's test results' details. [#17](https://github.com/iver-wharf/wharf-api/issues/17)
+Something like this,
+
+> ## Details | Logs | `Tests` :red_circle: | Artifacts
+>
+> - :green_circle: Passed: 256
+> - :yellow_circle: Skipped: 0
+> - :red_circle: **Failed: 3**
```

Comment:
nit: (purely cosmetic changes)

Need to [add `jemoji`](https://docs.github.com/en/enterprise/2.13/user/articles/emoji-on-github-pages) then it works *at least a little bit*:

![image](https://user-images.githubusercontent.com/2477952/124599064-8e931880-de65-11eb-8748-099a139cb0bf.png)

Seems like the https://www.webfx.com/tools/emoji-cheat-sheet/ is the source of truth here, and the green and yellow circle emojies are not yet added.

Suggest changing to something like:

```suggestion
> ## Details | Logs | `Tests` :x: | Artifacts
>
> - :heavy_check_mark: Passed: 256
> - :large_orange_diamond: Skipped: 0
> - :x: **Failed: 3**
```

![image](https://user-images.githubusercontent.com/2477952/124600664-3a893380-de67-11eb-910a-86676ab6bf32.png)

Reply:
Don't forget to add jemoji: https://docs.github.com/en/enterprise/2.13/user/articles/emoji-on-github-pages

Reply:
Added just now in https://github.com/iver-wharf/rfcs/pull/14/commits/ed272c479dc4024440b2e59b98121f8ea3012251